### PR TITLE
qcontainer: Support snapshots on blockdev mode for network storage

### PR DESCRIPTION
**Pull request by zhencliu**
ID: 1828693

Signed-off-by: Zhenchao Liu <zhencliu@redhat.com>

### CODE REVIEW

**Comment by zhencliu** [comment]
(1/1) Host_RHEL.m8.u2.product_av.iscsi_direct.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.rh_kernel_update: ERROR: 'list' object has no attribute 'splitlines' (235.14 s)
RESULTS    : PASS 0 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
(Known error, the error in 1828693 was gone)

 (1/5) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (2581.17 s)
 (2/5) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.rh_kernel_update: PASS (527.34 s)
 (3/5) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: CANCEL: Guest min memory has to be 4096.0, got 3072.0 (440.38 s)
 (4/5) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.netperf_stress_test.TCP_STREAM.guest2guest: PASS (262.21 s)
 (5/5) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.netperf_stress_test.TCP_STREAM.guest2guest: CANCEL: Guest min memory has to be 4096.0, got 3072.0 (30.25 s)

**Comment by zhencliu** [comment]
Hi @luckyh @YongxueHong , can you please help to review the tiny code change, thanks

File: virttest/qemu_devices/qcontainer.py [hunk 1]
```diff
@@ -2155,6 +2155,17 @@ def images_define_by_params(self, name, image_params, media=None,
...
+            # Use local image as the snapshot image
+            if image_params.get('enable_ceph') == 'yes':
+                sn_params['enable_ceph'] = 'no'
+            elif image_params.get('enable_iscsi') == 'yes':
+                sn_params['enable_iscsi'] = 'no'
+            elif image_params.get('enable_gluster') == 'yes':
+                sn_params['enable_gluster'] = 'no'
+            elif image_params.get('enable_nbd') == 'yes':
+                sn_params['enable_nbd'] = 'no'
+
```
Comment (YongxueHong) [inline virttest/qemu_devices/qcontainer.py:2168]: How about:
```
for name in ('ceph', 'iscsi', 'gluster', 'nbd'):
    if image_params.get('enable_%s' % name) == 'yes':
        sn_params['enable_%s' % name] = 'no'
        break
```

File: virttest/qemu_devices/qcontainer.py [hunk 2]
```diff
@@ -2165,7 +2176,10 @@ def images_define_by_params(self, name, image_params, media=None,
             self.temporary_image_snapshots.add(image_filename)
             image_encryption = storage.ImageEncryption.encryption_define_by_params(
                     sn, sn_params)
+            image_access = storage.ImageAccessInfo.access_info_define_by_params(
+                    sn, sn_params)
             imgfmt = 'qcow2'
+            name = sn  # now image is sn instead of image1
```
Comment (YongxueHong) [inline virttest/qemu_devices/qcontainer.py:2182]: I don't suggest to replace the `name` by `sn`. This just creates an offline snapshot to blockdev mode to surpport multiple vms which the cases boot them use the "snapshot = on" on drive mode.And the argument `name` maybe be used in the case level.

Comment (zhencliu) [reply virttest/qemu_devices/qcontainer.py:2182]: Hi Yongxue, 
given this situation, image1 is an iscsi image which needs secret object to access, sn is the local image, when forming the qemu-kvm command line, we need to associate the top image with the secret object if needed(pls refer to line 1713), here we define sn image with name=image1, then line 1713 will associate image1's secret object with sn image, which leads to unexpected error.

But your concern matters, two ways to solve this:
1. As only rh_kernel_update is affected for network storage backend, I can change rh_kernel_update that network storage is not supported.
2. set name=sn only when network storage backend is used.

What's your opinion, please?

File: virttest/qemu_devices/qcontainer.py (hunk 1 above)
Comment (zhencliu) [reply virttest/qemu_devices/qcontainer.py:2168]: OK, updated.

**Comment by zhencliu** [comment]
Close this request, for by now only rh_kernel_update is affected for the network storage backend, and we can disable network storage for it, which has been confirmed by the feature owner.

**Pull request closed without merging**